code/modules/urist/modules/GOAI/integrations/base_commander/mob_commander/obstacle_handlers/airlocks.dm

```diff
@@ -31,260 +39,565 @@
+		if(!airlock.p_open)
+			tracker.SetFailed()
+			DropObstacleMemory(airlock)
+			return
+
+	var/datum/wires/airlock_wires = airlock.wires
 
-	//VERY WIP. Fuck this AAA
-	//Code is unreachable as wip.
+	//Each wire attempt has a 5 second cooldown. So *max* this should take is 5*wire count, with some wriggle room for fixing wires + movement
+	if(TimedOutWalkDist(tracker, mob_pawn, airlock, (airlock_wires.wire_count * 5 SECONDS) + 10 SECONDS))
+		OBSTACLE_DEBUG_LOG("Hack task for [airlock] timed out!")
+		DropObstacleMemory(airlock)
+		tracker.SetFailed()
+		return
+
+	//We can't hack the airlock until we're next to it. Keep walking pal
+	if(!NavigateNextTo(tracker, mob_pawn, airlock))
+		return
 
-	//Todo: Add cooldown here
+	//5 second cooldown between hacks
+	var/last_action = tracker.BBGet("LastAction", null)
+	if(last_action && (world.time < last_action + 5 SECONDS))
+		return
 
-	//var/list/tried_wires = tracker.BBGet("WireAttempts", list())	//TBI
-	var/list/to_fix = tracker.BBGet("FixWires", list())
+	var/list/wire_queue = tracker.BBGet("WireQueue", list())
+	var/list/to_find = tracker.BBGet("TargetWires", wires.Copy())
+	var/list/tried_wires = tracker.BBGet("TriedWires", list())
 	var/list/known_wires = src.brain.GetMemoryValue(MEM_AIRLOCK_WIRES, list())
-	//var/datum/wires/airlock_wires = airlock.wires		//TBI
 
-	if(length(to_fix))
-		//var/wire = pick_n_take(to_fix) //TBI
-		tracker.BBSet("FixWires", to_fix)
-		//fix wire stuff here
+	//Someone's been here before us. Let's fix up everything first to give us a blank slate
+	if(!length(tried_wires) && airlock_wires.wires_status)
+		OBSTACLE_DEBUG_LOG("[airlock] previously hacked. Fixing wires.")
+		for(var/wire in airlock_wires.wires)
+			if(airlock_wires.IsColourCut(wire))
+				wire_queue[wire] = "mend"
+		tracker.BBSet("WireQueue", wire_queue)
+		return
+
+	//Process queued actions from the last tick
+	if(length(wire_queue))
+		var/wire = wire_queue[1]
+		var/action = wire_queue[wire]
+		var/is_cut = airlock_wires.IsColourCut(wire)
+		wire_queue -= wire
+		switch(action)
+			if("cut_wire")
+				OBSTACLE_DEBUG_LOG("Cutting wire [wire] for [airlock]")	//Temp debug logging
+				if(!is_cut)
+					airlock_wires.CutWireColour(wire)
+			if("cut_and_mend")
+				OBSTACLE_DEBUG_LOG("Cutting wire [wire] for [airlock]")
+				if(!is_cut)
+					airlock_wires.CutWireColour(wire)
+				wire_queue[wire] = "mend_wire"
+			if("mend_wire")
+				OBSTACLE_DEBUG_LOG("Mending wire [wire] for [airlock]")
+				if(is_cut)
+					airlock_wires.CutWireColour(wire)
+			if("pulse_wire")
+				OBSTACLE_DEBUG_LOG("Pulsing wire [wire] for [airlock]")
+				airlock_wires.PulseColour(wire)
+
+		tracker.BBSet("WireQueue", wire_queue)
+		tracker.BBSet("LastAction", world.time)
 		return
 
-	if(wires[1] in known_wires)
-		var/target_wire = wires[1]
-		wires.Remove(target_wire)
-		if(cut)
-			//Todo: cut stuff here
+	//We know what wire we want, queue it up
+	else if(to_find[1] in known_wires)
```

Comment:
Need to make sure this will not be out of bounds...

Reply:
Whoops. This is an artefact of where I moved branches around. Easy fix